streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1848,113 +1907,101 @@ object ZChannel {
+                               case Right(cause) =>
+                                 outgoing.offer(Exit.failCause(cause)) *> errorSignal.succeed(())
                              }
-                             .catchAllCause { cause =>
-                               queue.offer(Exit.failCause(cause)) *> errorSignal.succeed(()).unit
-                             }
-          _ <- pull
-                 .foldCauseZIO(
-                   cause =>
-                     queue.offer(Exit.failCause(cause)) *>
-                       ZIO.succeed(false),
-                   {
-                     case Left(outDone) =>
-                       errorSignal.await.interruptible.raceWith(permits.withPermits(n.toLong)(ZIO.unit).interruptible)(
-                         leftDone = (_, permitAcquisition) => permitAcquisition.interrupt.as(false),
-                         rightDone = (_, failureAwait) =>
-                           failureAwait.interrupt *>
-                             lastDone.get.flatMap {
-                               case Some(lastDone) => queue.offer(ZIO.succeed(Left(f(lastDone, outDone))))
-                               case None           => queue.offer(ZIO.succeed(Left(outDone)))
-                             }.as(false)
-                       )
-                     case Right(channel) =>
-                       mergeStrategy match {
+                           )
+        pullStrategy = mergeStrategy match {
                          case MergeStrategy.BackPressure =>
-                           for {
-                             latch <- Promise.make[Nothing, Unit]
-                             raceIOs =
-                               ZIO.scopedWith { scope =>
-                                 (queueReader >>> channel)
-                                   .toPullIn(scope)
-                                   .flatMap(evaluatePull(_).raceAwait(errorSignal.await.interruptible))
-                               }
-                             childFiber <- permits
-                                             .withPermit(latch.succeed(()) *> raceIOs)
-                                             .forkIn(scope)
-                             _       <- latch.await
-                             errored <- errorSignal.isDone
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               latch <- Promise.make[Nothing, Unit]
+                               raceIOs =
+                                 ZIO.scopedWith { scope =>
+                                   (incoming >>> channel)
+                                     .toPullInAlt(scope)
+                                     .flatMap(evaluatePull)
+                                 }
+                               _ <- permits
+                                      .withPermit(latch.succeed(()) *> raceIOs)
+                                      .fork
```

Comment:
Interesting, but these worker fibers are short-lived, aren't they removed from the set after they're done?

Reply:
So this is where it gets a bit tricky; they're removed from the set once we no longer hold any reference to them (since they're stored in a `WeakHashMap`). If for some the `FiberRuntime` keeps being referenced after completion, it will remain in the children set which might cause it to grow indefinitely.

Having said that (and thinking about it a bit more), if some part of the application is causing fibers to be indefinitely referenced, then memory leak is going to occur regardless. So I think using `fork` probably isn't as big of an issue

Reply:
Yeah, if that were to happen that would be flagged as a bug either in the runtime or in the layer above that uses the stream API. IMHO there is very little reason to create a scope within a scope, I only ever use them to escape the implicit scope of the parent fiber.